pkg/kv/kvserver/gc/gc.go

```diff
@@ -390,140 +457,481 @@ func processReplicatedKeyRange(
+// If later non garbage data is found, batcher looks on number of keys covered
+// by already checked range and will either:
+//   - send clear range batch up to the previous key
+//   - restore state from checkpoint and send point key batch then rewind to the
+//     version after the last sent key
+type gcKeyBatcher struct {
+	gcKeyBatcherThresholds
+
+	prevWasNewest bool
+	// Points batches are accumulated until we cross clear range threshold.
+	// There's always at least one points batch initialized.
+	pointsBatches []pointsBatch
+	// totalMemUsed sum of memUsed of all points batches cached to avoid
+	// recomputing on every check.
+	totalMemUsed int
+
+	// Tracking of clear range requests.
+	clearRangeEndKey          roachpb.Key
+	clearRangeStartKey        storage.MVCCKey
+	clearRangeCounters        gcBatchCounters
+	partialPointBatchCounters gcBatchCounters
+
+	gcer PureGCer
+}
+
+func (b *gcKeyBatcher) String() string {
+	pbs := make([]string, len(b.pointsBatches))
+	for i, pb := range b.pointsBatches {
+		pbs[i] = pb.String()
+	}
+	return fmt.Sprintf("gcKeyBatcher{batches=[%s], clearRangeCnt=%s, partialBatchCnt=%s, totalMem=%d}",
+		strings.Join(pbs, ", "),
+		b.clearRangeCounters.String(),
+		b.partialPointBatchCounters.String(),
+		b.totalMemUsed)
+}
+
+func (b *gcKeyBatcher) foundNonGCableData(
+	ctx context.Context, cur *mvccKeyValue, isNewestPoint bool,
+) (counterUpdate gcInfoUpdate, err error) {
+	b.prevWasNewest = isNewestPoint
+	if !b.clearRangeEnabled {
+		return counterUpdate, nil
+	}
+
+	// Check if there are any complete clear range or point batches collected
+	// and flush them as we reached end of current consecutive key span.
+	counterUpdate, err = b.maybeFlushPendingBatches(ctx)
+	if err != nil {
+		return gcInfoUpdate{}, err
 	}
+	b.clearRangeCounters = gcBatchCounters{}
 
-	// Iterate all versions of all keys from oldest to newest. If a version is an
-	// intent it will have the highest timestamp of any versions and will be
-	// followed by a metadata entry. The loop will determine whether a given key
-	// has garbage and, if so, will determine the timestamp of the latest version
-	// which is garbage to be added to the current batch. If the current version
-	// pushes the size of keys to be removed above the limit, the current key will
-	// be added with that version and the batch will be sent. When the newest
-	// version for a key has been reached, if haveGarbageForThisKey, we'll add the
-	// current key to the batch with the gcTimestampForThisKey.
-	var (
-		batchGCKeys           []roachpb.GCRequest_GCKey
-		batchGCKeysBytes      int64
-		haveGarbageForThisKey bool
-		gcTimestampForThisKey hlc.Timestamp
-		sentBatchForThisKey   bool
-	)
-	it := makeGCIterator(desc, snap, threshold, excludeUserKeySpan)
-	defer it.close()
-	for ; ; it.step() {
-		s, ok := it.state()
-		if !ok {
-			if it.err != nil {
-				return false, it.err
+	if isNewestPoint {
+		b.clearRangeEndKey = append(b.clearRangeEndKey[:0], cur.key.Key...)
```

Comment:
It shouldn't affect correctness, but we could nil it out when flushing batch if we are going to clear counters there. I thought that it was easier to understand if clearing of state would be done outside of flush because it is already pretty overloaded with sending data out and accounting for it. But if it is easier to understand like that let it be.

Reply:
Thanks. Seems better for the flush to update any associated state, so that callers don't have to remember doing it (potentially getting it wrong).